pygeos/strtree.py

```diff
@@ -186,3 +189,72 @@ def query_bulk(self, geometry, predicate=None):
             predicate = BinaryPredicate[predicate].value
 
         return self._tree.query_bulk(geometry, predicate)
+
+    @requires_geos("3.6.0")
+    def nearest(self, geometry, max_distance=None, return_distance=False):
+        """Returns the index of the nearest item in the tree for each input
+        geometry.
+
+        If there are multiple equidistant or intersected geometries in tree, all
+        should be returned.  Tree indexes are returned in the order they are visited
+        for each input geometry and may not be in ascending index order; no meaningful
+        order is implied.
+
+        The max_distance used to search for nearest items in the tree may have a
+        significant impact on performance by reducing the number of input geometries
+        that are evaluated for nearest items in the tree.  Only those input geometries
+        with at least one tree item within +/- max_distance beyond their envelope will
+        be evaluated.
+
+        The distance, if returned, will be 0 for any intersected geometries in the tree.
+
+        Any geometry that is None or empty in the input geometries is omitted from
+        the output.
+
+        Parameters
+        ----------
+        geometry : Geometry or array_like
+            Input geometries to query the tree.
+        max_distance : float, optional (default: None)
+            Maximum distance within which to query for nearest items in tree.  By default
+            will return nearest tree item for each input geometry regardless of distance.
+            If provided, only those pairs of input geometry and nearest tree item within
+            this distance will be included in the results.  0 and None are equivalent.
```

Comment:
I was trying to avoid adding additional parsing boilerplate in C that would be required if we passed down `None` to that level.

I think we should steer users toward `query_bulk(arr, predicate='intersects') to get intersecting geometries from the tree, so there is not benefit to allowing them to set a `0` value here, and we can keep that as a sentinel value that `max_distance` is not set.  I've added additional validation checks to make sure if it is not `None` that it is > 0.